base/complex.jl

```diff
@@ -1028,3 +1028,68 @@ function complex(A::AbstractArray{T}) where T
     end
     convert(AbstractArray{typeof(complex(zero(T)))}, A)
 end
+
+function div(a::Complex{T}, b::Complex{V}, r::RoundingMode=RoundNearest) where {T<:Integer, V<:Integer}
+    R = promote_type(T, V)
+    a, b = Complex{R}(a), Complex{R}(b)
+    b̅ = conj(b)
+    # TODO: Handle overflow when calculating a*b̅
+    t = a*b̅
```

Comment:
Since `div` can't overflow for real numbers, I think, we should try to guarantee that `div` doesn't overflow for complex numbers either. I think this safety is worth the extra performance cost in most uses of this function. Implementing `Base.checked_mul(::Complex{<:Integer}, ::Complex{<:Integer})` should actually be fairly straightforward to do.

Reply:
That's true, but an issue with the current division algorithm (a/b = a*b̅/abs2(b)), is that due to abs2 overflow, we cannot ensure that a/b *never overflows*. We can indeed try to catch and handle all the possible overflows, but that will be it.

We can always find some inputs which have b::Complex{T} and abs2(b)>typemax(T), which will make div(a, b) overflow.

E.g. a = Complex{Int32}(1, 1), b= Complex{Int32}(10^5, 10^5). Here, a/b will overflow due to the current algorithm.

Reply:
I'm not sure, what you mean. I mean, we check in `abs2` and in `*`, that nothing overflows and then you're just calling `div` on real integers, which can't overflow, so we should catch all cases of overflow. I'm wondering though, if there aren't any better algorithms for division of complex integers, because checking in each step will prevent overflow in the result, but we will also throw errors for cases where either `a*b'` or `abs2(b)` overflows, but `a/b` should actually still be representable as the complex integer type of `a` and `b`.

Reply:
> we will also throw errors for cases where either `a*b'` or `abs2(b)` overflows

This is what I meant. It is possible that abs2(b) overflows, but a/b is still representable as a complex integer. E.g. `div(Complex{Int32}(1, 1), Complex{Int32}(10^5, 10^5))`.

Reply:
As soon as one of `abs(b)` or `a*b'` overflows, isn't the end result also wrong with your current algorithm, even if the result might still be representable?

Reply:
> Also, should checked_mul(::Complex, ::Complex) be in complex.jl or in checked.jl ?
complex.jl is probably better. Defining it in checked.jl might even cause problems during bootstrap, because I think that module gets loaded before complex.jl

Reply:
Yes. Probably we need to find a better algorithm if there exists one. The best we can do right now is catching all the overflows.

Reply:
Just an idea: how about multiplying by `(1 / abs(b))^2` instead? Not sure what that will due to precision problems though. But something along those lines might work.

Reply:
If you already have an approximate result there may be a way to improve it using a suitable Newton method.

Reply:
So you think this should internally use floating point arithmetic? That might not be a bad idea here.

Reply:
Or use rationals? Or promote to a wider type? To be honest I don't quite understand what's going on in the PR and what the constraints are.

Reply:
@ravibitsgoa Could you explain your usecase for this? That would probably help to understand some of the constraints

Reply:
https://github.com/JuliaLang/julia/issues/35125
This PR is based on this issue. After quite a long discussion there, we decided we should avoid using floating points or rationals or wider type internally. Although those three are quite decent solutions IMO, but they will create a problem when we want to find exact integer answers of div, rem and gcd.

Reply:
I really cannot imagine those overflow-checked stuff (if you get it right) will ever be more efficient than just calculating with widened precision.

Reply:
Can someone help me find good test cases for performance testing for these methods?
Since I am new to Julia and do not have much experience with performance testing. I am unable to construct tests for which gcd would take more iterations. 
I tried running a few tests for the gcd method, which take like 0.000001 seconds to complete.

Reply:
For Euclidean GCD algorithm, the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Algorithmic_efficiency) occurs when arguments are consecutive Fibonacci numbers.

Using this, I tested gcd(f1 + 0im, f2+ 0im), for f1 = 1836311903, f2= 2971215073, since these are the largest Int64 arguments for which gcd does not overflow.

Benchmarking with @time gives `0.000009 seconds (3 allocations: 96 bytes)` for the current overflow-checked implementation.

I compared it against another implementation, which calculates gcd by widening arguments into `Complex{BigInt}` first. For such a complex gcd method, @time gives around `
  0.000150 seconds (2.23 k allocations: 46.305 KiB)`. 

Hence, we can safely conclude that for arguments below up to f1, f2, calculating gcd without widening into `Complex{BigInt}` might be a better decision.

If someone finds better benchmarks, I am happy to try them as well.